receiver/k8sclusterreceiver/collection/metadata.go

```diff
@@ -25,42 +25,158 @@ import (
+	resourceID string
+	// metadata is a set of key-value pairs that describe a resource.
+	metadata map[string]string
 }
 
 // getGenericMetadata is responsible for collecting metadata from K8s resources that
 // live on v1.ObjectMeta.
 func getGenericMetadata(om *v1.ObjectMeta, resourceType string) *KubernetesMetadata {
 	rType := strings.ToLower(resourceType)
-	properties := utils.MergeStringMaps(map[string]string{}, om.Labels)
+	metadata := utils.MergeStringMaps(map[string]string{}, om.Labels)
 
-	properties[k8sKeyWorkLoadKind] = rType
-	properties[k8sKeyWorkLoadName] = om.Name
-	properties[fmt.Sprintf("%s.creation_timestamp",
+	metadata[k8sKeyWorkLoadKind] = rType
+	metadata[k8sKeyWorkLoadName] = om.Name
+	metadata[fmt.Sprintf("%s.creation_timestamp",
 		rType)] = om.GetCreationTimestamp().Format(time.RFC3339)
 
 	for _, or := range om.OwnerReferences {
-		properties[strings.ToLower(or.Kind)] = or.Name
-		properties[strings.ToLower(or.Kind)+"_uid"] = string(or.UID)
+		metadata[strings.ToLower(or.Kind)] = or.Name
+		metadata[strings.ToLower(or.Kind)+"_uid"] = string(or.UID)
 	}
 
 	return &KubernetesMetadata{
 		resourceIDKey: getResourceIDKey(rType),
 		resourceID:    string(om.UID),
-		properties:    properties,
+		metadata:      metadata,
 	}
 }
 
 func getResourceIDKey(rType string) string {
 	return fmt.Sprintf("k8s.%s.uid", rType)
 }
+
+// mergeKubernetesMetadataMaps merges maps of string (resource id) to
+// KubernetesMetadata into a single map.
+func mergeKubernetesMetadataMaps(maps ...map[string]*KubernetesMetadata) map[string]*KubernetesMetadata {
+	out := map[string]*KubernetesMetadata{}
+	for _, m := range maps {
+		for id, km := range m {
+			out[id] = km
+		}
+	}
+
+	return out
+}
+
+// KubernetesMetadataExporter provides an interface to implement
+// ConsumeKubernetesMetadata in Exporters that support metadata.
+type KubernetesMetadataExporter interface {
+	// ConsumeKubernetesMetadata will be invoked every time there's an
+	// update to a resource that results in one or more KubernetesMetadataUpdate.
+	ConsumeKubernetesMetadata(metadata []*KubernetesMetadataUpdate) error
+}
+
+// KubernetesMetadataUpdate provides a delta view of metadata on a resource between
+// two revisions of a resource.
+type KubernetesMetadataUpdate struct {
+	// ResourceIDKey is the label key of UID label for the resource.
+	ResourceIDKey string
+	// ResourceID is the Kubernetes UID of the resource. In case of
+	// containers, this value is the container id.
+	ResourceID string
+	// MetadataToAdd contains key-value pairs that are newly added to
+	// the resource description in the current revision.
+	MetadataToAdd map[string]string
+	// MetadataToUpdate contains key-value pairs that have been updated
+	// in the current revision compared to the previous revisions(s).
+	MetadataToUpdate map[string]string
+	// MetadataToRemove contains key-value pairs that no longer describe
+	// a resource and needs to be removed.
+	MetadataToRemove map[string]string
```

Comment:
What is the "key" in these maps? Is it possible to define type alias to make it clear?

Reply:
Added an example in the comments. Let me know if that makes it clear what the keys in these maps are. If not, we could maybe add a type called `metadataKey`?

Reply:
Nice examples.